common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+  // present bit tag on it so that this layer can manage tagging the high bit of
+  // a present byte.
+  CARBON_DCHECK((present_byte & PresentMask) == 0)
+      << llvm::formatv("{0:x}", present_byte);
+
+  // Use a simple fallback approach for sizes beyond 8.
+  // TODO: Instead of a silly fallback, we should generalize the below
+  // algorithm for sizes above 8, even if to just exercise the same code on
+  // more platforms.
+  if constexpr (Size > 8) {
+    static_assert(Size <= 32, "Sizes larger than 32 not yet supported!");
+    uint32_t mask = 0;
+    uint32_t bit = 1;
+    present_byte |= PresentMask;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] == present_byte) {
+        mask |= bit;
+      }
+      bit <<= 1;
+    }
+    return MatchRange(mask);
+  }
+
+  // This algorithm only works for matching *present* bytes. We leverage the
+  // set high bit in the present case as part of the algorithm. The whole
+  // algorithm has a critical path height of 4 operations, and does 6
+  // operations total on AArch64. The operation dependency graph is:
+  //
+  //          group | MSBs    LSBs * match_byte + MSBs
+  //                 \            /
+  //                 mask ^ pattern
+  //                      |
+  // group & MSBs    MSBs - mask
+  //        \            /
+  //    group_MSBs & mask
+  //
+  // This diagram and the operation count are specific to AArch64 where we have
+  // a fused *integer* multiply-add operation.
+  //
+  // While it is superficially similar to the "find zero bytes in a word" bit
+  // math trick, it is different because this is designed to have no false
+  // positives and perfectly produce 0x80 for matching bytes and 0x00 for
+  // non-matching bytes. This is do-able because we constrain to only handle
+  // present matches which only require testing 7 bits and have a particular
+  // layout.
+
+  // Set the high bit of every byte to `1`. The match byte always has this bit
+  // set as well, which ensures the xor below, in addition to zeroing the byte
+  // that matches, also clears the high bit of every byte.
+  uint64_t mask = byte_ints[0] | MSBs;
+  // Broadcast the match byte to all bytes, and mask in the present bits in the
+  // MSBs of each byte. We structure this as a multiply and an add because we
+  // know that the add cannot carry, and this way it can be lowered using
+  // combined multiply-add instructions if available.
+  uint64_t pattern = LSBs * present_byte + MSBs;
+  CARBON_DCHECK(pattern == (LSBs * present_byte | MSBs))
+      << "Unexpected carry from addition!";
+
+  // Xor the broadcast pattern, making matched bytes become zero bytes.
+  mask = mask ^ pattern;
+  // Subtract the mask bytes from `0x80` bytes so that any non-zero mask byte
+  // clears the high bit but zero leaves it intact.
+  mask = MSBs - mask;
+  // Mask down to the high bits, but only those in the original group.
```

Comment:
I feel like you could be a bit more explicit about what is happening here.
```suggestion
  // Zero everything but the high bits, and also zero the high bits of any bytes for "not present" slots in the original group. This avoids false positives for `Empty` and `Deleted` bytes in the metadata.
```